test/Conversion/FIRRTLToHW/lower-to-hw.mlir

```diff
@@ -1674,4 +1315,283 @@ firrtl.circuit "Simple"   attributes {annotations = [{class =
+  firrtl.module @ConnectNestedFieldsAndIndices(in %clock: !firrtl.clock,
+                                 in %reset: !firrtl.uint<1>,
+                                 in %value: !firrtl.uint<2>,
+                                 out %result: !firrtl.vector<bundle<a: vector<bundle<b: uint<2>>, 3>>, 5>) {
+    %count = firrtl.reg %clock: !firrtl.vector<bundle<a: vector<bundle<b: uint<2>>, 3>>, 5>
+    %field1 = firrtl.subindex %count[1] : !firrtl.vector<bundle<a: vector<bundle<b: uint<2>>, 3>>, 5>
+    %field2 = firrtl.subfield %field1(0) : (!firrtl.bundle<a: vector<bundle<b: uint<2>>, 3>>) -> !firrtl.vector<bundle<b: uint<2>>, 3>
+    %field3 = firrtl.subindex %field2[1] : !firrtl.vector<bundle<b: uint<2>>, 3>
+    %field4 = firrtl.subfield %field3(0) : (!firrtl.bundle<b: uint<2>>) -> !firrtl.uint<2>
+    firrtl.strictconnect %field4, %value : !firrtl.uint<2>
+
+    // CHECK: %count = seq.firreg %17 clock %clock : !hw.array<5xstruct<a: !hw.array<3xstruct<b: i2>>>>
+    // CHECK-DAG: %1 = hw.array_get %count[%c1_i3] : !hw.array<5xstruct<a: !hw.array<3xstruct<b: i2>>>>
+    // CHECK-DAG: %2 = hw.struct_extract %1["a"] : !hw.struct<a: !hw.array<3xstruct<b: i2>>>
+    // CHECK-DAG: %3 = hw.array_get %2[%c1_i2] : !hw.array<3xstruct<b: i2>>
+    // CHECK-DAG: %4 = hw.struct_extract %3["b"] : !hw.struct<b: i2>
+    // CHECK-DAG: %5 = hw.array_get %count[%c1_i3_0] : !hw.array<5xstruct<a: !hw.array<3xstruct<b: i2>>>>
+    // CHECK-DAG: %6 = hw.struct_extract %5["a"] : !hw.struct<a: !hw.array<3xstruct<b: i2>>>
+    // CHECK-DAG: %7 = hw.array_get %6[%c1_i2_1] : !hw.array<3xstruct<b: i2>>
+    // CHECK-DAG: %8 = hw.struct_inject %7["b"], %value : !hw.struct<b: i2>
+    // CHECK-DAG: %9 = hw.array_slice %6[%c0_i2] : (!hw.array<3xstruct<b: i2>>) -> !hw.array<1xstruct<b: i2>>
+    // CHECK-DAG: %10 = hw.array_create %8 : !hw.struct<b: i2>
+    // CHECK-DAG: %11 = hw.array_slice %6[%c-2_i2] : (!hw.array<3xstruct<b: i2>>) -> !hw.array<1xstruct<b: i2>>
+    // CHECK-DAG: %12 = hw.array_concat %9, %10, %11 : !hw.array<1xstruct<b: i2>>, !hw.array<1xstruct<b: i2>>, !hw.array<1xstruct<b: i2>>
+    // CHECK-DAG: %13 = hw.struct_inject %5["a"], %12 : !hw.struct<a: !hw.array<3xstruct<b: i2>>>
+    // CHECK-DAG: %14 = hw.array_slice %count[%c0_i3] : (!hw.array<5xstruct<a: !hw.array<3xstruct<b: i2>>>>) -> !hw.array<1xstruct<a: !hw.array<3xstruct<b: i2>>>>
+    // CHECK-DAG: %15 = hw.array_create %13 : !hw.struct<a: !hw.array<3xstruct<b: i2>>>
+    // CHECK-DAG: %16 = hw.array_slice %count[%c2_i3] : (!hw.array<5xstruct<a: !hw.array<3xstruct<b: i2>>>>) -> !hw.array<3xstruct<a: !hw.array<3xstruct<b: i2>>>>
+    // CHECK-DAG: %17 = hw.array_concat %14, %15, %16 : !hw.array<1xstruct<a: !hw.array<3xstruct<b: i2>>>>, !hw.array<1xstruct<a: !hw.array<3xstruct<b: i2>>>>, !hw.array<3xstruct<a: !hw.array<3xstruct<b: i2>>>>
+  }
```

Comment:
This and some other tests are quite long. Can these be simplified using aggregates with fewer elements to improve the future maintainability of the tests without compromising the coverage of the tests?

Can these be slimmed down to a smaller number of targeted tests and can the existing tests be pruned to be more pointed in what they are testing?

Also, I like `CHECK-DAG`. However, is there a good reason to be using it here. I would expect that there should be predictability of the lowering as opposed to "these lowering in any order". (Pedantically, even though these use `CHECK-DAG`, that isn't helping due to the use of fixed SSA names. 😉 )

Reply:
Individual aspects of the lowering are tested already by shorter tests, this is a big one to check if they combine correctly. I used `CHECK-DAG` since there are constants interleaved at various points preventing the use of `CHECK-NEXT`.

Reply:
(Nit) It seems better to just use `CHECK` then. I read this, seeing `CHECK-DAG`, as "these checks can appear in any order in the output" when they're being used, instead, to circumvent interleaved constants. I.e., aren't all of these just `CHECK`s?